File: galaxy-rtp-validator-service/src/main/java/com/finzly/galaxy/rtp/validator/service/PaymentTemplateService.java
package com.finzly.galaxy.rtp.validator.service;

import com.finzly.galaxy.rtp.validator.dto.PaymentInfoRequest;
import com.finzly.galaxy.rtp.validator.dto.PaymentTemplate;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

@Service
@Slf4j
public class PaymentTemplateService {
    
    // In-memory storage for templates (replace with database in production)
    private final ConcurrentHashMap<Long, PaymentTemplate> templates = new ConcurrentHashMap<>();
    private final AtomicLong idGenerator = new AtomicLong(1);

    public PaymentTemplate saveTemplate(PaymentInfoRequest request) {
        log.info("Saving payment template: {}", request.getTemplateName());
        
        PaymentTemplate template = PaymentTemplate.builder()
                .id(idGenerator.getAndIncrement())
                .templateName(request.getTemplateName())
                .senderRoutingNumber(request.getSenderRoutingNumber())
                .receiverRoutingNumber(request.getReceiverRoutingNumber())
                .senderAccountNumber(request.getSenderAccountNumber())
                .receiverAccountNumber(request.getReceiverAccountNumber())
                .amount(request.getAmount())
                .currency(request.getCurrency())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();
        
        templates.put(template.getId(), template);
        log.info("Template saved successfully with ID: {}", template.getId());
        
        return template;
    }

    public List<PaymentTemplate> getAllTemplates() {
        log.info("Retrieving all payment templates. Count: {}", templates.size());
        return new ArrayList<>(templates.values());
    }

    public PaymentTemplate getTemplateById(Long id) {
        log.info("Retrieving template by ID: {}", id);
        return templates.get(id);
    }

    public PaymentTemplate updateTemplate(Long id, PaymentInfoRequest request) {
        log.info("Updating template ID: {}", id);
        
        PaymentTemplate existing = templates.get(id);
        if (existing == null) {
            log.warn("Template not found with ID: {}", id);
            return null;
        }
        
        existing.setTemplateName(request.getTemplateName());
        existing.setSenderRoutingNumber(request.getSenderRoutingNumber());
        existing.setReceiverRoutingNumber(request.getReceiverRoutingNumber());
        existing.setSenderAccountNumber(request.getSenderAccountNumber());
        existing.setReceiverAccountNumber(request.getReceiverAccountNumber());
        existing.setAmount(request.getAmount());
        existing.setCurrency(request.getCurrency());
        existing.setUpdatedAt(LocalDateTime.now());
        
        log.info("Template updated successfully");
        return existing;
    }

    public boolean deleteTemplate(Long id) {
        log.info("Deleting template ID: {}", id);
        PaymentTemplate removed = templates.remove(id);
        return removed != null;
    }

    public long getTemplateCount() {
        return templates.size();
    }
}

File: galaxy-rtp-validator-ui/src/app/services/payment-template.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PaymentInfoRequest, PaymentTemplate } from '../models/payment.model';

@Injectable({
  providedIn: 'root'
})
export class PaymentTemplateService {
  private apiUrl = 'http://localhost:8080/api/payment-templates';

  constructor(private http: HttpClient) { }

  saveTemplate(request: PaymentInfoRequest): Observable<PaymentTemplate> {
    return this.http.post<PaymentTemplate>(this.apiUrl, request);
  }

  getAllTemplates(): Observable<PaymentTemplate[]> {
    return this.http.get<PaymentTemplate[]>(this.apiUrl);
  }

  getTemplateById(id: number): Observable<PaymentTemplate> {
    return this.http.get<PaymentTemplate>(`${this.apiUrl}/${id}`);
  }

  updateTemplate(id: number, request: PaymentInfoRequest): Observable<PaymentTemplate> {
    return this.http.put<PaymentTemplate>(`${this.apiUrl}/${id}`, request);
  }

  deleteTemplate(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getTemplateCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/count`);
  }
}

